src/node/http.ts

```diff
@@ -1,885 +1,253 @@
+  logger.debug(`redirecting from ${req.originalUrl} to ${redirectPath}`)
+  res.redirect(redirectPath)
 }
 
 /**
- * An HTTP server. Its main role is to route incoming HTTP requests to the
- * appropriate provider for that endpoint then write out the response. It also
- * covers some common use cases like redirects and caching.
+ * Get the value that should be used for setting a cookie domain. This will
+ * allow the user to authenticate only once. This will use the highest level
+ * domain (e.g. `coder.com` over `test.coder.com` if both are specified).
  */
-export class HttpServer {
-  protected readonly server: http.Server | https.Server
-  private listenPromise: Promise<string | null> | undefined
-  public readonly protocol: "http" | "https"
-  private readonly providers = new Map<string, HttpProvider>()
-  public readonly heart: Heart
-  private readonly socketProvider = new SocketProxyProvider()
-
-  /**
-   * Provides the actual proxying functionality.
-   */
-  private readonly proxy = proxy.createProxyServer({})
-
-  public constructor(private readonly options: HttpServerOptions) {
-    this.heart = new Heart(path.join(paths.data, "heartbeat"), async () => {
-      const connections = await this.getConnections()
-      logger.trace(plural(connections, `${connections} active connection`))
-      return connections !== 0
-    })
-    this.protocol = this.options.cert ? "https" : "http"
-    if (this.protocol === "https") {
-      this.server = httpolyglot.createServer(
-        {
-          cert: this.options.cert && fs.readFileSync(this.options.cert),
-          key: this.options.certKey && fs.readFileSync(this.options.certKey),
-        },
-        this.onRequest,
-      )
-    } else {
-      this.server = http.createServer(this.onRequest)
-    }
-    this.proxy.on("error", (error, _request, response) => {
-      response.writeHead(HttpCode.ServerError)
-      response.end(error.message)
-    })
-    // Intercept the response to rewrite absolute redirects against the base path.
-    this.proxy.on("proxyRes", (response, request: ProxyRequest) => {
-      if (response.headers.location && response.headers.location.startsWith("/") && request.base) {
-        response.headers.location = request.base + response.headers.location
+export const getCookieDomain = (host: string, proxyDomains: string[]): string | undefined => {
+  const idx = host.lastIndexOf(":")
+  host = idx !== -1 ? host.substring(0, idx) : host
+  if (
+    // Might be blank/missing, so there's nothing more to do.
+    !host ||
+    // IP addresses can't have subdomains so there's no value in setting the
+    // domain for them. Assume anything with a : is ipv6 (valid domain name
+    // characters are alphanumeric or dashes).
+    host.includes(":") ||
+    // Assume anything entirely numbers and dots is ipv4 (currently tlds
+    // cannot be entirely numbers).
+    !/[^0-9.]/.test(host) ||
+    // localhost subdomains don't seem to work at all (browser bug?).
```

Comment:
Oof yeah this comment did not make the problem clear. With localhost domains the cookies don't seem to filter down. So if my domain is `dev.localhost` and I go to `8080.dev.localhost` this code would try to set the cookie at `dev.localhost` but it can't be read at `8080.dev.localhost` which means you can never log in there.

We could instead set a cookie at `8080.dev.localhost` and `dev.localhost`, maybe that's what you were thinking with the multiple cookie question? That seems decent to me although I suppose if we ever implement a logout it could be a bit tricky getting all the cookies since we won't know which sub-domains they've been set for. (I think?)